src/systems/sprites.js

```diff
@@ -187,224 +177,234 @@ function createGeometry(maxSprites) {
-            el.classList.add("ui");
-            scene.appendChild(el);
-            el.setObject3D("mesh", mesh);
-            mesh.frustumCulled = false;
-            mesh.renderOrder = window.APP.RENDER_ORDER.HUD_ICONS;
-            mesh.raycast = this.raycast.bind(this);
-          }
-        );
+  constructor(scene) {
+    this.sprites = { action: [], notice: [] };
+    this.indicesFromSprites = { action: new WeakMap(), notice: new WeakMap() };
+    this.meshes = { action: null, notice: null };
+
+    waitForDOMContentLoaded().then(() => {
+      for (const type in PNGS) {
+        const spritesheetPng = PNGS[type];
+        createImageTexture(spritesheetPng, getThemeColorShifter(type)).then(spritesheetTexture => {
+          const geometry = createGeometry(MAX_SPRITES);
+          const material = createMaterial(spritesheetTexture);
+          const mesh = (this.meshes[type] = new THREE.Mesh(geometry, material));
+          const el = document.createElement("a-entity");
+          el.classList.add("ui");
+          scene.appendChild(el);
+          el.setObject3D("mesh", mesh);
+          mesh.frustumCulled = false;
+          mesh.renderOrder = window.APP.RENDER_ORDER.HUD_ICONS;
+          mesh.raycast = this.raycast.bind(this);
+        });
       }
     });
 
-    APP.store.addEventListener("themechanged", async () => {
-      for (const [spritesheetPng, type] of pngs) {
+    APP.store.addEventListener("themechanged", () => {
+      for (const type in PNGS) {
+        const spritesheetPng = PNGS[type];
+        // TODO: Fix me if possible
+        // 1. If theme is changed before meshes are initialized, no effect to textures.
+        // 2. If theme is changed before textures loading for the previous theme is completed
+        // and textures loading for the new theme is completed earlier than the previous ones,
+        // the previous ones are set.
         if (this.meshes[type]) {
-          const newTexture = await createImageTexture(spritesheetPng, getThemeColorShifter(type));
-          const oldTexture = this.meshes[type].material.uniforms.u_spritesheet.value;
-          if (oldTexture) {
+          createImageTexture(spritesheetPng, getThemeColorShifter(type)).then(newTexture => {
+            const oldTexture = this.meshes[type].material.uniforms.u_spritesheet.value;
+            this.meshes[type].material.uniforms.u_spritesheet.value = newTexture;
+            this.meshes[type].material.uniformsNeedUpdate = true;
             disposeTexture(oldTexture);
-          }
-          this.meshes[type].material.uniforms.u_spritesheet.value = newTexture;
-          this.meshes[type].material.uniformsNeedUpdate = true;
+          });
         }
       }
     });
   }
 
   tick(t) {
-    if (!this.meshes.action || !this.meshes.notice) return;
-
-    for (let i = 0; i < SHEET_TYPES.length; i++) {
-      const sheetType = SHEET_TYPES[i];
+    for (const sheetType of SHEET_TYPES) {
```

Comment:
No clue if this is still the case, but at one point we observed these sort of loops created garbage so we tried to avoid them in hot code. Maybe just use a regular for just in case?

Reply:
> these sort of loops

Does it mean `for (const foo of array)`?

Reply:
Yep. for-of and for-in loops use iterators and may create garbage. Like I said not sure if this is still true and in which engines, but it was true at some point.

Reply:
OK, using old for loop sounds safer. Updated, thanks.

And it's good if it's written in the best practice document or somewhere else.